src/borda_clj/client.clj

```diff
@@ -16,30 +17,57 @@
    buffer and stops the background thread that does the submitting."
+    [max-buffer-size interval send on-send-error]
+    (let [measurements  (atom {})
+          running       (atom true)
+          submit        (fn [dimensions values] (swap! measurements collect max-buffer-size dimensions values))
+          resubmit      (fn [m] (doseq [[dimensions values] m] (submit dimensions values)))
+          flush         (fn [on-flush-error]
+                          (let [[m] (reset-vals! measurements {})]
+                            (if (not-empty m)
+                              (try
+                                (send m)
+                                (catch Throwable e
+                                  (on-flush-error m e))))))
+          stop          (fn [] (reset! running false) (flush on-send-error))]
+                                        ; periodically send to borda
+      (future (while @running (do
+                                (Thread/sleep interval)
+                                (flush (fn [m e] (on-send-error m e) (resubmit m))))))
+      [submit stop])))
+
+(defn reducing-submitter
   [max-buffer-size interval send on-send-error]
-  (let [measurements  (atom {})
-        running       (atom true)
-        submit        (fn [dimensions values] (swap! measurements collect max-buffer-size dimensions values))
-        resubmit      (fn [m] (doseq [[dimensions values] m] (submit dimensions values)))
-        flush         (fn [on-flush-error]
-                        (let [[m] (reset-vals! measurements {})]
-                          (if (not-empty m)
-                            (try
-                              (send m)
-                              (catch Throwable e
-                                (on-flush-error m e))))))
-        stop          (fn [] (reset! running false) (flush on-send-error))]
-        ; periodically send to borda
-    (future (while @running (do
-                              (Thread/sleep interval)
-                              (flush (fn [m e] (on-send-error m e) (resubmit m))))))
-    [submit stop]))
+  (let [report-ch (chan)]
+    (go-loop [measurements {}
+              flush-timeout (timeout interval)]
+      (let [flush (fn [on-error]
+                    (when (not-empty measurements)
+                      (try
+                        (send measurements)
+                        (catch Throwable e
+                          (on-error measurements e)))))
+            [[dimensions values :as v] ch] (alts! [report-ch flush-timeout])]
+          (cond (= ch flush-timeout)
+                (do
+                  (flush (fn [m e]
+                           (on-send-error m e)
+                           (onto-chan report-ch m false)))
+                  (recur {} (timeout interval)))
+                (some? v)
+                (recur (collect measurements max-buffer-size dimensions values)
+                       flush-timeout)
```

Comment:
Reading from a closed Clojure channel yields `nil`, which is otherwise not a legal value to send through a channel.